--- a/src/usr/local/www/widgets/widgets/log.widget.php
+++ b/src/usr/local/www/widgets/widgets/log.widget.php
@@ -122,6 +122,25 @@
        </thead>
        <tbody>
 <?php
+
+       // https://redmine.pfsense.org/issues/12673
+       $cmdtoexecute="/sbin/pfctl -vvPsr | grep ^@";
+       $thefile = shell_exec($cmdtoexecute);
+       $rule_lines = explode("\n", $thefile, -1);
+       $idx = 0;
+       $norules = 0;
+       foreach ($rule_lines as $someline) {
+               $firstcha = substr($someline, 0, 1);
+               if ($firstcha == "@") {
+                       $rulekeys[] = array(
+                               'rulenum' => substr($someline, 1, strpos($someline, ' ')),
+                               'rawidx'  => $idx,
+                       );
+                       $norules++;
+               }
+               $idx++;
+       }
+
        foreach ($filterlog as $filterent):
                if ($filterent['version'] == '6') {
                        $srcIP = "[" . htmlspecialchars($filterent['srcip']) . "]";
@@ -141,7 +160,14 @@
                        $iconfn = "check text-success";
                }

-               $rule = find_rule_by_number($filterent['rulenum'], $filterent['tracker'], $filterent['act']);
+               $rule = "no rule info available";
+               foreach ($rulekeys as $actrule) {
+                       if ($actrule['rulenum'] == $filterent['rulenum']) {
+                               $rawidx = $actrule['rawidx'];
+                               $rule = $rule_lines[$rawidx];
+                               break;
+                       }
+               }

                // Putting <wbr> tags after each ':'  allows the string to word-wrap at that point
                $srcIP = str_replace(':', ':<wbr>', $srcIP);
