--- a/src/usr/local/www/includes/functions.inc.php
+++ b/src/usr/local/www/includes/functions.inc.php
@@ -133,10 +133,10 @@
 }

 function get_mbuf(&$mbuf, &$mbufpercent) {
-       $mbufs_output=trim(`/usr/bin/netstat -mb | /usr/bin/grep "mbuf clusters in use" | /usr/bin/awk '{ print $1 }'`);
-       list($mbufs_current, $mbufs_cache, $mbufs_total, $mbufs_max) = explode("/", $mbufs_output);
-       $mbuf = "{$mbufs_total}/{$mbufs_max}";
-       $mbufpercent = ($mbufs_max > 0) ? round(($mbufs_total / $mbufs_max) * 100, 0) : "NA";
+        $mbufs_total = get_single_sysctl("vm.uma.mbuf_cluster.limit.items");
+        $mbufs_max = get_single_sysctl("vm.uma.mbuf_cluster.limit.max_items");
+        $mbuf = "{$mbufs_total}/{$mbufs_max}";
+        $mbufpercent = ($mbufs_max > 0) ? round(($mbufs_total / $mbufs_max) * 100, 0) : "NA";
 }

 function get_temp() {
@@ -423,9 +423,9 @@
 }

 function get_load_average() {
-       $load_average = "";
-       exec("/usr/bin/uptime | /usr/bin/sed 's/^.*: //'", $load_average);
-       return $load_average[0];
+        $cleaned = trim(get_single_sysctl("vm.loadavg"), "{} ");
+        $result = str_replace(" ", ", ", $cleaned);
+        return $result;
 }

 function get_interfacestats() {
