--- a/src/etc/inc/util.inc
+++ b/src/etc/inc/util.inc
@@ -54,21 +54,26 @@
 }
 
 function isvalidpid($pidfile) {
-       $output = "";
-       if (file_exists($pidfile)) {
-               exec("/bin/pgrep -qnF {$pidfile} 2>/dev/null", $output, $retval);
-               return (intval($retval) == 0);
-       }
-       return false;
+       if (file_exists($pidfile)) {
+               $pid = (int)trim(file_get_contents($pidfile));
+               return ($pid > 0 && posix_kill($pid, 0));
+       }
+        return false;
 }
 
 function is_process_running($process) {
-       $output = "";
-       if (!empty($process)) {
-               exec("/bin/pgrep -anx " . escapeshellarg($process), $output, $retval);
-               return (intval($retval) == 0);
-       }
-       return false;
+        static $data = null; 
+       if (!empty($process)) {
+                if ($data === null) {
+                       $cmd = "ps aux --libxo json";
+                       $data = json_decode(shell_exec($cmd));
+               } 
+                foreach ($data->{'process-information'}->process as $p) {
+                        if (stripos($p->command, $process) !== false)
+                                return true;
+                }
+        }
+        return false;
 }
 
 function isvalidproc($proc) {
@@ -2359,29 +2364,33 @@
  * Get values of sysctl OID's listed in $names (accepts an array or a single
  * name) and return an array of key/value pairs set for those that exist
  */
-function get_sysctl($names) {
+function get_sysctl($names)
+{
+               static $data = null;
+               if ($data === null) {
+                $data = new stdClass();
+                $cmd="sysctl dev hw kern machdep vm.pmap vm.uma.pf_states vm.uma.mbuf_cluster vm.stats";
+                $output=shell_exec($cmd);
+                $array = explode("\n", trim($output));
+                foreach ($array as $line) {
+                        list($key, $value) = explode(": ", $line);
+                        $data->$key = $value;
+                       }
+               }
+
        if (empty($names)) {
                return array();
        }
 
-       if (is_array($names)) {
-               $name_list = array();
-               foreach ($names as $name) {
-                       $name_list[] = escapeshellarg($name);
-               }
-       } else {
-               $name_list = array(escapeshellarg($names));
-       }
-
-       exec("/sbin/sysctl -iq " . implode(" ", $name_list), $output);
        $values = array();
-       foreach ($output as $line) {
-               $line = explode(": ", $line, 2);
-               if (count($line) == 2) {
-                       $values[$line[0]] = $line[1];
+               if (!is_array($names)) {
+               $names = [$names];
+               }
+               foreach ($names as $name) {
+                if (property_exists($data, $name)) {
+                        $values[$name] = (is_numeric($data->{$name})) ? (int) $data->{$name} : $data->{$name};
                }
        }
-
        return $values;
 }

@@ -2862,29 +2871,31 @@

 /* Return system's route table */
 function route_table() {
-       exec("/usr/bin/netstat --libxo json -nWr", $rawdata, $rc);
+        static $data = null;
+        if ($data === null) {
+                $rawdata = [];
+                exec("/usr/bin/netstat --libxo json -nWr", $rawdata, $rc);
+                if ($rc != 0) {
+                        return array();
+                }
+                $data = json_decode(implode(" ", $rawdata), JSON_OBJECT_AS_ARRAY);
+        }

-       if ($rc != 0) {
-               return array();
-       }
+        $netstatarr = $data['statistics']['route-information']['route-table']['rt-family'];

-       $netstatarr = json_decode(implode(" ", $rawdata), JSON_OBJECT_AS_ARRAY);
-       $netstatarr = $netstatarr['statistics']['route-information']
-           ['route-table']['rt-family'];
+        $result = array();
+        $result['inet'] = array();
+        $result['inet6'] = array();
+        foreach ($netstatarr as $item) {
+                if ($item['address-family'] == 'Internet') {
+                        $result['inet'] = $item['rt-entry'];
+                } else if ($item['address-family'] == 'Internet6') {
+                        $result['inet6'] = $item['rt-entry'];
+                }
+        }
+        unset($netstatarr);

-       $result = array();
-       $result['inet'] = array();
-       $result['inet6'] = array();
-       foreach ($netstatarr as $item) {
-               if ($item['address-family'] == 'Internet') {
-                       $result['inet'] = $item['rt-entry'];
-               } else if ($item['address-family'] == 'Internet6') {
-                       $result['inet6'] = $item['rt-entry'];
-               }
-       }
-       unset($netstatarr);
-
-       return $result;
+        return $result;
 }

 /**
@@ -2983,7 +2994,7 @@
                return '';
        }

-       $route = route_get('default', $ipprotocol, true);
+       $route = route_get('default', $ipprotocol, false);

        if (empty($route)) {
                return '';
